docs/RFCS/leaseholder_rebalancing.md

```diff
@@ -0,0 +1,136 @@
+difference between leaseholders and followers. These extra duties
+impose additional load on the leaseholder making it desirable to
+spread that load throughout a cluster in order to improve performance.
+
+The last motivation is to place the leaseholder for a range near the
+gateway node that is accessing the range in order to minimize network
+RTT. As an obvious example: it is preferable for the leaseholder to be
+in the same datacenter as the gateway node.
+
+# Detailed design
+
+This RFC is intended to address the first two motivations and punt on
+the last one (load-based leaseholder placement). Note that addressing
+the problem of evenly distributing leaseholders across a cluster will
+also address the inability to rebalance a replica away from the
+leaseholder as we'll always have sufficient non-leaseholder replicas
+in order to perform rebalancing.
+
+Leaseholder rebalancing will be performed using a similar mechanism to
+replica rebalancing. The periodically gossiped `StoreCapacity` proto
+will be extended with a `leaseholder_count` field. Currently,
+`replicateQueue` has the following logic:
+
+1. If range has dead replicas, remove them.
+2. If range is under-replicated, add a replica.
+3. If range is over-replicated, remove a replica.
+4. If the range needs rebalancing, add a replica.
+
+The proposal is to add the following logic (after the above replica
+rebalancing logic):
+
+5. If the leaseholder is on an overfull store transfer the lease to
+the least loaded follower less than the mean.
+6. If the leaseholder store has a leaseholder count above the mean and
+one of the followers has an underfull leaseholder count transfer the
+lease to the least loaded follower.
+
+# Testing
+
+Individual rebalancing heuristics can be unit tested, but seeing how
+those heuristics interact with a real cluster can often reveal
+surprising behavior. We have an existing allocation simulation
+framework, but it has seen infrequent use. As an alternative, the
+`zerosum` tool has been useful in examining rebalancing heuristics. We
+propose to fork `zerosum` and create a new `allocsim` tool which will
+create a local N-node cluster with controls for generating load and
+using smaller range sizes to test various rebalancing scenarios.
+
+# Future Directions
+
+We eventually need to provide load-based leaseholder placement, both
+to place leaseholders close to gateway nodes and to more accurately
+balance load across a cluster. Balancing load by balancing replica
+counts or leaseholder counts does not capture differences in per-range
+activity. For example, one table might be significantly more active
+than others in the system making it desirable to distribute the ranges
+in that table more evenly.
+
+At a high-level, expected load on a node is proportional to the number
+of replicas/leaseholders on the node. But a more accurate description
+is that it is proportional to the number of bytes on the node. Rather
```

Comment:
Which part do you think we're already doing? Range splits can definitely trigger rebalancing for exactly the reason mentioned in this paragraph. See #9435 for more details.